go-controller/pkg/cluster/node.go

```diff
@@ -108,6 +109,76 @@ func (cluster *OvnClusterController) StartClusterNode(name string) error {
 	defer f.Close()
 	confJSON := fmt.Sprintf("{\"name\":\"ovn-kubernetes\", \"type\":\"%s\"}", config.CNI.Plugin)
 	_, err = f.Write([]byte(confJSON))
+	if err != nil {
+		return err
+	}
+
+	err = cluster.watchNamespaceUpdate(node, subnet.String())
+	return err
+}
+
+// If default namespace MasterOverlayIP annotation has been chaged, update
+// config.OvnNorth and config.OvnSouth auth with new ovn-nb and ovn-remote
+// IP address
+func (cluster *OvnClusterController) updateOvnNode(masterIP string,
+	node *kapi.Node, subnet string) error {
+	err := config.UpdateOvnNodeAuth(masterIP)
+	if err != nil {
+		return err
+	}
+	err = setupOVNNode(node.Name, config.Kubernetes.APIServer,
+		config.Kubernetes.Token, config.Kubernetes.CACert)
+	if err != nil {
+		return err
+	}
+
+	err = util.RestartOvnController(config.OvnSouth.ClientAuth)
+	if err != nil {
+		return err
+	}
+
+	// Recreate logical switch and management port for this node
+	err = ovn.CreateManagementPort(node.Name, subnet,
+		cluster.ClusterIPNet.String(),
+		cluster.ClusterServicesSubnet)
+	if err != nil {
+		return err
+	}
+
+	// Reinit Gateway for this node if the --init-gateways flag is set
+	if cluster.GatewayInit {
+		if runtime.GOOS == windowsOS {
+			panic("Windows is not supported as a gateway node")
+		}
+		err = cluster.initGateway(node.Name, cluster.ClusterIPNet.String(),
+			subnet)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
 
+// watchNamespaceUpdate starts watching namespace resources and calls back
+// the update handler logic if there is any namspace update event
+func (cluster *OvnClusterController) watchNamespaceUpdate(node *kapi.Node,
+	subnet string) error {
+	_, err := cluster.watchFactory.AddNamespaceHandler(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(old, newer interface{}) {
+				//oldNs := old.(*kapi.Namespace)
+				//oldMasterIP := oldNs.Annotations[MasterOverlayIP]
+				newNs := newer.(*kapi.Namespace)
+				newMasterIP := newNs.Annotations[MasterOverlayIP]
```

Comment:
Looks like we will call an update whenever a namespace changes instead of just a change in IP.

Reply:
Done, updated to check if newMasterIP != oldMasterIP